src/modules/schema.js

```diff
@@ -0,0 +1,189 @@
+import { stringToHex, u8aToString, u8aToHex } from '@polkadot/util';
+import { validate } from 'jsonschema';
+import axios from 'axios';
+
+import { getHexIdentifierFromDID } from '../utils/did';
+import { getSignatureFromKeyringPair } from '../utils/misc';
+import { isHexWithGivenByteSize } from '../utils/codec';
+import Signature from '../signatures/signature';
+
+import {
+  DockBlobIdByteSize,
+  blobHexIdToQualified,
+  createNewDockBlobId,
+  getHexIdentifierFromBlobID,
+} from './blob';
+
+// Supported schemas
+import JSONSchema07 from '../utils/vc/schemas/schema-draft-07';
+
+export default class Schema {
+  /**
+   * Creates a new `Schema` object
+   * @constructor
+   * @param {string} [id] - optional schema ID, if not given, generate a random id
+   */
+  constructor(id) {
+    this.id = id || createNewDockBlobId();
+    this.name = '';
+    this.version = '1.0.0';
+  }
+
+  /**
+   * Add the JSON schema to this object after checking that `json` is a valid JSON schema. Check if JSON is valid.
+   * @param {object} json - the schema JSON
+   */
+  async setJSONSchema(json) {
+    await Schema.validateSchema(json);
+    this.schema = json;
+    return this;
+  }
+
+  setName(name) {
+    this.name = name;
+    return this;
+  }
+
+  /**
+   * Update the object with `author` key. Repeatedly calling it will keep resetting the author
+   * did can be a DID hex identifier or full DID
+   * @param {string} did - the author DID
+   */
+  setAuthor(did) {
+    if (did.startsWith('did:') || isHexWithGivenByteSize(did, DockBlobIdByteSize)) {
+      this.author = did;
+    } else {
+      throw new Error(`Supplied author ${did} is not a valid DID or ${DockBlobIdByteSize} byte hex string`);
+    }
+    return this;
+  }
+
+  /**
+   * Update the object with `signature` key. This method is used when
+   * signing key/capability is not present and the signature is received from outside.
+   * Repeatedly calling it will keep resetting the `signature` key.
+   * The signature must be one of the supported objects
+   * @param {object} signature - The schema's signature
+   */
+  setSignature(signature) {
+    if (signature instanceof Signature) {
+      this.signature = signature;
+    } else {
+      throw new Error('Provided signature object is not of instance Signature');
+    }
+
+    return this;
+  }
+
+  /**
+   * Serializes the object using `getSerializedBlob` and then signs it using the given
+   * polkadot-js pair. The object will be updated with key `signature`. Repeatedly calling it will
+   * keep resetting the `signature` key
+   * @param {any} msg - The message to sign
+   * @param {object} pair - Key pair to sign with
+   */
+  sign(msg, pair) {
```

Comment:
Why does it take a message to sign? The spec said to use the object state

Reply:
"temporary" implementation (you can see things calling sign passing [1,2,3,4] just to confirm it works) because blob serialization method wasnt implemented when this was. remains a todo, should've had a comment

Reply:
Ok

Reply:
resolved, however sign method needed to take additional parameter for blob module